keymanapp/keyman
# change(web): touch alias optimization followup

Continues from #6551, as a review for it came in after it was already merged.

Makes one mild feature/tweak:  users may now touch a point _slightly_ after the proper location for a caret without having the caret jump to after the following character.  (See:  "snap leniency".)

Makes one niche fix:  attempts to place the caret significantly after the end of a line could cause the caret to be placed after the first character of the following line.  Now, it reliably follows the end-of-line space if it exists.

One last major distinction worth noting:  upon revisiting the code part referred to by https://github.com/keymanapp/keyman/pull/6551#discussion_r857188115, I realized that there was little reason we couldn't resolve that complexity by simply relocating the code to within the `TouchAliasElement` class, facilitating removal of the indirection noted in that comment.  Should've followed through with _that_ idea in #6551, as I'd actually thought of moving it (briefly) even then.  

(Additionally, analysis of the function showed that it didn't even refer to any values specific to its original owning class and could actually even _drop_ parameters this way, as the `TouchAliasElement` was, itself, one of them and contains a field holding another of them!)

Beyond that... I addressed most of those post-merge comments, cleaned up the code, and clarified certain execution conditions.  I currently plan to leave the arrow-function changes to #6557.

Part of me feels like the Y-axis code and X-axis code are kinda WET... but DRYing it out completely would likely add a bit much complexity.  I mean, four separate binary search implementations within the same function really can't be considered DRY, even if each has its own nuances.

# User Testing

## Tests

Use remote device emulation on Google Chrome on the `web/testing/issue5312-touch-alias-optimization` testing page, listed on the testing index page near the bottom as "Tests performance of touch-alias elements with large amounts of text (#5312)".

- GROUP_PHONE:  Run tests on the page while emulating an iPhone SE via Chrome mobile-device emulation.
   - Note that the page will likely auto-scroll on your first touch within the testing text.  (The smaller resolution forces this.)
- GROUP_TABLET:  Run tests on the page while emulating an iPad Air (or similar tablet form-factor) using Chrome mobile-device emulation.

- TEST_INITIAL_TOUCH:  

    - Touch text in the middle of the visible paragraph, preferably toward the center vertically and horizontally.
    - The caret should land very near to the touched location, practically under the touch if possible.
    - Repeat at least 3 times in different locations; ensure it behaves as expected.
    - Finally, repeat once more slightly to the left or right of the previous repetition; ensure it behaves as expected.

- TEST_SCROLLED_TOUCH:

    - Scroll the textbox down three paragraphs.
    - Touch text in the middle of the now-top (fourth overall) paragraph, preferably toward the center vertically and horizontally.
    - The caret should land very near to the touched location, practically under the touch if possible.
    - Repeat at least 3 times in different locations; ensure it behaves as expected.
    - Finally, repeat once more slightly to the left or right of the previous repetition; ensure it behaves as expected.

- TEST_LINE_FINAL_TOUCH:
    - Attempt to place the caret at the end of a "short" line in the middle of a paragraph, touching far to the right within the text area.
    - The caret should be placed at the end of the line, after a space.

        <details><summary>Image for comparison</summary>
       
        <img width="320" alt="image" src="https://user-images.githubusercontent.com/25213402/165139122-44934402-744a-48a4-94c0-a2fd5ea3acde.png">
        
        Note how there is plenty of space to the right of the caret - try touching _there_.
        
        </details>
